src/ultk/effcomm/information_bottleneck/tools.py

```diff
@@ -0,0 +1,134 @@
+import numpy as np
+from scipy.special import softmax
+from scipy.stats import entropy
+
+##############################################################################
+# Numerical precision helpers
+##############################################################################
+
+PRECISION = 1e-15
+
+
+def get_gaussian_noise(shape):
+    """Small Gaussian noise."""
+    return np.random.normal(loc=0.0, scale=1e-15, size=shape)
+
+
+def add_noise_to_stochastic_matrix(q, weight=1e-2):
+    """
+    Given an input stochastic matrix `q`, sample a stochastic matrix `p` and
+    mix it with the input with a small weight `weight`, i.e., return q + weight * p.
+    """
+    # Generate a stochastic matrix `p` using a Dirichlet distribution
+    p = np.random.dirichlet(np.ones(q.shape[1]), size=q.shape[0])
+    # Mix `q` with `p` using the specified weight
+    noisy_matrix = q + weight * p
+    # Normalize to ensure the rows sum to 1
+    noisy_matrix /= noisy_matrix.sum(axis=1, keepdims=True)
+    return noisy_matrix
+
+
+def random_stochastic_matrix(shape, gamma=1e-10):
+    """
+    Initialize a stochastic matrix (2D array) that sums to 1 along the rows.
+
+    Args:
+        shape: tuple, the desired shape of the stochastic matrix (e.g., `(rows, cols)`).
+        gamma: float, scaling factor for the random normal initialization.
+
+    Returns:
+        A stochastic matrix with rows summing to 1.
+    """
+    energies = gamma * np.random.randn(*shape)
+    return softmax(energies, axis=1)
+
+
+##############################################################################
+# Probability and Information
```

Comment:
What do you think about putting this stuff in the probability sub-module?

Reply:
Yes. I hope one day we move probability out of effcomm too